Add functionality to output the UUID for log lookup purposes

daoauth opened the pull request:
# Summary

Add functionality to [output the UUID](https://github.com/slsa-framework/slsa-github-generator/issues/3741) for log lookup purposes. The `provenance-rekor-uuid` parameter represents the Rekor UUID, a unique identifier that can be used to search for and view specific log entry details on the Rekor Search UI.

## Testing Process

Forked the branch and conducted unit tests to verify the output of the UUID.

## Checklist

- [x] Review the contributing [guidelines](https://github.com/slsa-framework/slsa-github-generator/blob/main/CONTRIBUTING.md)
- [x] Add a reference to related issues in the PR description.
- [x] Update documentation if applicable.
- [ ] Add unit tests if applicable.
- [ ] Add changes to the [CHANGELOG](https://github.com/slsa-framework/slsa-github-generator/blob/main/CHANGELOG.md) if applicable.

laurentsimon commented on .github/workflows/generator_generic_slsa3.yml:
```diff
@@ -97,6 +97,9 @@ on:
       provenance-name:
         description: "The artifact name of the signed provenance. (A file with the intoto.jsonl extension)."
         value: ${{ jobs.generator.outputs.provenance-name }}
+      provenance-uuid:
```
nit: Let's call it `provenance-rekor-uuid`

daoauth replied:
Thank you for the feedback. I have made the requested changes.

laurentsimon commented on .github/workflows/generator_generic_slsa3.yml:
```diff
@@ -97,6 +97,9 @@ on:
         description: "The artifact name of the signed provenance. (A file with the intoto.jsonl extension)."
         value: ${{ jobs.generator.outputs.provenance-name }}
+      provenance-uuid:
+        description: "You could also use the Rekor Search UI: https://search.sigstore.dev/?uuid= to view the log entry details."
```
nit: Describe what the output is: `The rekor UUID ... It may be used to search ...`

Also, don't forget to update the README https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/generic#workflow-outputs

daoauth replied:
Thank you for the feedback. I have made the requested changes

ramonpetgrave64 commented:
Thanks for implementing. Please use the PR template in this PR's description.

daoauth commented:
> Thanks for implementing. Please use the PR template in this PR's description.

Thank you for the feedback. I have made the requested changes

ramonpetgrave64 commented:
Thanks for finding a way to test, can you show us an example run in your fork?

ramonpetgrave64 commented on internal/builders/generic/README.md:
```diff
@@ -277,6 +277,7 @@ The [generic workflow](https://github.com/slsa-framework/slsa-github-generator/b
 | Name              | Description                                                                                     |
 | ----------------- | ----------------------------------------------------------------------------------------------- |
 | `provenance-name` | The artifact name of the signed provenance.                                                     |
+| `provenance-rekor-uuid` | The Rekor UUID is a unique identifier that can be used to search for and view specific log entry details on the Rekor Search UI.|
```
reformat the table, please

daoauth replied:
I have updated the table as requested.

daoauth commented:
> Thanks for finding a way to test, can you show us an example run in your fork?

I renamed `slsa-framework/slsa-github-generator/` to `zktx-io/slsa-github-generator/` and created a release. Additionally, I downloaded `slsa-generator-generic-linux-amd64`, added it to the release, and modified the code to execute it as follows.

``` bash
      - name: Create and sign provenance
        id: sign-prov
        continue-on-error: true
        # NOTE: Inputs and `github` context are set to environment variables in
        # order to avoid script injection.
        # See: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections
        env:
          GITHUB_CONTEXT: "${{ toJSON(github) }}"
          VARS_CONTEXT: "${{ toJSON(vars) }}"
          UNTRUSTED_PROVENANCE_NAME: "${{ inputs.provenance-name }}"
        run: |
          set -euo pipefail

          # I downloaded the files uploaded to the release and changed their attributes using chmod +x to make them executable.
          chmod +x "$GITHUB_WORKSPACE/$BUILDER_BINARY"
          untrusted_prov_name=""
          if [ "$UNTRUSTED_PROVENANCE_NAME" != "" ]; then
            untrusted_prov_name="$UNTRUSTED_PROVENANCE_NAME"
          fi
          # Create and sign provenance.
          # NOTE: The builder verifies that the provenance path is located
          # in the current directory.
          # NOTE: $untrusted_provenance_path may be empty. In this case the
          # attest command chooses a file name based on the subject name and
          # number of subjects based on in-toto attestation bundle file naming conventions.
          # See: https://github.com/in-toto/attestation/blob/main/spec/bundle.md#file-naming-convention
          # NOTE: The attest commmand outputs the provenance-name and provenance-sha256
          output=$("$GITHUB_WORKSPACE/$BUILDER_BINARY" attest --subjects-filename "${SUBJECTS_FILENAME}" -g "$untrusted_prov_name")
          echo "$output"
          uuid=$(echo "$output" | grep -oP 'UUID \K[0-9a-f]{64}')

          # I verified the output by printing the UUID.
          echo "$uuid"
          echo "uuid=$uuid" >> $GITHUB_OUTPUT
```

daoauth commented:
> Sorry for not clarifying. If you ran the workflow on your own fork, can you provide us with a link so we can inspect the logs?

1. This is the test repository where I forked slsa-github-generator: https://github.com/zktx-io/slsa-github-generator-test
2. This is the test repository for outputting logs: https://github.com/zktx-io/slsa-github-generator-test2/actions/runs/10091019235/job/27901935588

<img width="1243" alt="Screenshot 2024-07-25 at 5 52 47 PM" src="https://github.com/user-attachments/assets/6648f4a9-2eb3-46f5-9fe3-4669f0aecde1">

ramonpetgrave64 commented on .github/workflows/generator_generic_slsa3.yml:
```diff
@@ -234,7 +238,10 @@ jobs:
-          "$GITHUB_WORKSPACE/$BUILDER_BINARY" attest --subjects-filename "${SUBJECTS_FILENAME}" -g "$untrusted_prov_name"
+          output=$("$GITHUB_WORKSPACE/$BUILDER_BINARY" attest --subjects-filename "${SUBJECTS_FILENAME}" -g "$untrusted_prov_name")
+          echo "$output"
+          uuid=$(echo "$output" | grep -oP 'UUID \K[0-9a-f]{80}')
```
nonblocker: it seems the rekor UUID is meant to be 64 hex digits
 - https://github.com/sigstore/rekor/blob/5e502df5603712e553ab4a077793bdece66d4dbc/pkg/api/error.go#L42

daoauth replied:
The length of the UUID generated in "create and sign provenance" is 80.

https://search.sigstore.dev/?uuid=24296fb24b8ad77a037b175b6a98db3d318ae8d94428f5be89eb07c8a6b70b58b54bdd32c7b15fb6

https://docs.sigstore.dev/logging/sharding/#identifier-definitions-entryid-uuid-logid-log-index

So, I think the UUID used as a search parameter in sigstore is actually an entry ID.

haydentherapper replied:
The additional prefix digits are for distinguishing which log "shard" hosts the content.

ramonpetgrave64 commented:
@haydentherapper please take a look

ramonpetgrave64 reviewed (commented):
Looks alright, just add the details to the list of unreleased changes.
- https://github.com/slsa-framework/slsa-github-generator/blob/main/CHANGELOG.md#unreleased-vars-context-recorded-in-provenance

ramonpetgrave64 commented on CHANGELOG.md:
```diff
@@ -112,6 +113,11 @@ duplication."
+### Unreleased: Provenance Rekor UUID Output
+
+- **Added**: The workflow now includes the output of `provenance-rekor-uuid` for log lookup purposes.
+  This enhancement ensures that each build and deployment can be traced and audited using a unique identifier.
```
Builds can already be traced, as the are recorded in the Sigstore transparency logs. slsa-verifier also performs the lookups during its verification.

```suggestion
  This enhancement ensures that each build and deployment can be traced and audited using a unique identifier.
```

ramonpetgrave64 commented on CHANGELOG.md:
```diff
@@ -112,6 +113,11 @@ duplication."
+### Unreleased: Provenance Rekor UUID Output
+
+- **Added**: The workflow now includes the output of `provenance-rekor-uuid` for log lookup purposes.
+  This enhancement ensures that the provenance-rekor-uuid can be used to search for and view specific log entry details on the Rekor Search UI.
```
```suggestion
  The provenance-rekor-uuid can be used to search for and view specific log entry details on the Rekor Search UI.
```

ramonpetgrave64 commented on .github/workflows/generator_generic_slsa3.yml:
```diff
@@ -234,7 +238,10 @@ jobs:
+          output=$("$GITHUB_WORKSPACE/$BUILDER_BINARY" attest --subjects-filename "${SUBJECTS_FILENAME}" -g "$untrusted_prov_name")
+          echo "$output"
+          uuid=$(echo "$output" | grep -oP 'UUID \K[0-9a-f]{80}')
+          echo "uuid=$uuid" >> $GITHUB_OUTPUT
```
Lint errors: please run `make actionlint` and `make lint`

ramonpetgrave64 commented:
Closing due to inactivity.
